File: most-beautiful-item-for-each-query/most-beautiful-item-for-each-query.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<int> maximumBeauty(vector<vector<int>>& items, vector<int>& queries) {
        // Step 1: Sort items by the first value (beauty)
        sort(items.begin(), items.end());

        // Step 2: Precompute the prefix maximum beauty values
        int n = items.size();
        vector<int> prefix_max(n);
        prefix_max[0] = items[0][1];
        for (int i = 1; i < n; i++) {
            prefix_max[i] = max(prefix_max[i - 1], items[i][1]);
        }

        // Step 3: Process each query using binary search
        vector<int> answer;
        for (int q : queries) {
            // Use binary search to find the largest index where items[i][0] <= q
            // `upper_bound` gives us the first element that is greater than the query
            int idx = upper_bound(items.begin(), items.end(), vector<int>{q, INT_MAX}) - items.begin() - 1;
            if (idx == -1) {
                answer.push_back(0);  // No valid item found
            } else {
                answer.push_back(prefix_max[idx]);
            }
        }
        return answer;
    }
};

int main() {
    Solution sol;
    vector<vector<int>> items = {{4, 6}, {7, 5}, {2, 10}, {3, 8}};
    vector<int> queries = {5, 10, 7};
    
    vector<int> result = sol.maximumBeauty(items, queries);
    
    for (int val : result) {
        cout << val << " ";
    }
    cout << endl;

    return 0;
}

File: most-beautiful-item-for-each-query/most-beautiful-item-for-each-query.py
from typing import List
from bisect import bisect_right

class Solution:
    def maximumBeauty(self, items: List[List[int]], queries: List[int]) -> List[int]:
        # Step 1: Sort items by the first value (beauty).
        items.sort(key=lambda x: x[0])
        
        # Step 2: Precompute the prefix maximum beauty values.
        prefix_max = []
        max_beauty = 0
        for item in items:
            max_beauty = max(max_beauty, item[1])
            prefix_max.append(max_beauty)
        
        # Step 3: For each query, use binary search to find the rightmost item.
        answer = []
        for query in queries:
            # bisect_right finds the index where the query could be inserted.
            idx = bisect_right(items, [query, float('inf')]) - 1
            
            # If idx is -1, no items satisfy the condition
            if idx == -1:
                answer.append(0)
            else:
                answer.append(prefix_max[idx])
        
        return answer
         
    
def main():
    items = [[1,2],[3,2],[2,4],[5,6],[3,5]]
    queries = [1,2,3,4,5,6]
    soltion = Solution()
    result = soltion.maximumBeauty(items, queries)
    print(result)

if __name__ == "__main__":
    main()